packages/compose-tunnel-agent/src/api-server/ws/handlers/exec.ts

```diff
@@ -0,0 +1,62 @@
+import { inspect } from 'util'
+import { createWebSocketStream } from 'ws'
+import { parseQueryParams, queryParamBoolean } from '../../query-params'
+import { wsHandler } from '../handler'
+import { NotFoundError } from '../../http-server-helpers'
+
+const handler = wsHandler(
+  /^\/container\/([^/?]+)\/exec($|\?)/,
+  async (ws, req, match, { log, docker, dockerFilter }) => {
+    const containerId = match[1]
+    if (!await dockerFilter.inspectContainer(containerId)) {
+      throw new NotFoundError()
+    }
+    const { obj: { tty: ttyQueryParam }, search } = parseQueryParams(req.url ?? '', { tty: true })
```

Comment:
It might worth splitting it to two ops in the future:
One for creating the exec/logs handler (return handle url), the other for reading in websocket based on the handle.
The handle can up-for-grabs and available in a limited time window. (can be saved even in CTA memory, or localfs)

It can be a bit safer as POST requests can work with authorization header and CORS.